File: Frontend/src/types/message.ts
// Mirrors a single translated utterance, whether it originates from the
// sign-detection pipeline, the speech-to-text pipeline, or a manually
// selected quick phrase.
export type MessageSource = 'sign' | 'speech' | 'phrase'

// One score per recognition/translation pipeline stage. Which keys show
// up depends on the message's source - a 'sign' message never carries
// speechRecognition, a 'speech' message never carries handSignRecognition -
// never invent a score for a stage that didn't run on that message.
export type ConfidenceCategory = 'handSignRecognition' | 'speechRecognition' | 'translation'
export type ConfidenceBreakdown = Partial<Record<ConfidenceCategory, number>>

export interface ConversationMessage {
  id: string
  sessionId: string
  source: MessageSource
  text: string
  timestamp: string // ISO 8601
  // 0-100, from the ML model. Absent for manually-selected phrases, which
  // were never run through detection - never fake a number here. When
  // confidenceBreakdown is present, this is that breakdown's average -
  // kept alongside it so existing average-only views (Live Conversation,
  // the History list) don't need to know about categories at all.
  confidence?: number
  // Per-stage detail behind `confidence` above. Absent wherever
  // `confidence` itself is absent (phrases).
  confidenceBreakdown?: ConfidenceBreakdown
}

// The envelope the future WebSocket stream will emit. Kept separate from
// ConversationMessage so transport-level fields never leak into UI state.
export interface TranslationSocketEvent {
  type: 'translation' | 'status' | 'error'
  payload: ConversationMessage | DeviceStatusEvent | { message: string }
}

export interface DeviceStatusEvent {
  camera?: string
  hands?: string
  face?: string
  microphone?: string
  speaker?: string
  ai?: string
  lightLevel?: string
}

File: Backend/app/models/message.py
from typing import Literal
from pydantic import BaseModel

MessageSource = Literal["sign", "speech"]


# Mirrors src/types/message.ts -> ConversationMessage.
class ConversationMessage(BaseModel):
    id: str
    sessionId: str
    source: MessageSource
    text: str
    timestamp: str  # ISO 8601
    confidence: int  # 0-100, from the ML model


# Mirrors src/types/device.ts -> DeviceStatus, used as the "status" event
# payload over the WebSocket (see ws/translate.py).
class DeviceStatusEvent(BaseModel):
    camera: str | None = None
    hands: str | None = None
    face: str | None = None
    microphone: str | None = None
    speaker: str | None = None
    ai: str | None = None
    lightLevel: str | None = None


class ErrorPayload(BaseModel):
    message: str


# Mirrors src/types/message.ts -> TranslationSocketEvent. `payload` is
# deliberately loose here (frontend narrows on `type`) - FastAPI/Pydantic
# will serialize whichever concrete payload object is passed in.
class TranslationSocketEvent(BaseModel):
    type: Literal["translation", "status", "error"]
    payload: ConversationMessage | DeviceStatusEvent | ErrorPayload


# Mirrors the shape of control messages the frontend's RealTranslationSocket
# sends via socket.send(JSON.stringify({ type: 'control', action })).
class ControlMessage(BaseModel):
    type: Literal["control"]
    action: str  # "end-session" | "export" today; kept as str so new
    # actions (e.g. a future "submit-speech") don't require a backend
    # schema change.
